### ISSUE TITLE
Code navigation accessibility: Support cursor-driven (text caret) keyboard navigation

### ISSUE THREAD

**Issue by umpox** [sourcegraph/sourcegraph#42420]
## Description

We should allow configuring a cursor for full accessibility in the blob view.

**Acceptance criteria**:
- A setting is added that allows enabling cursor-driven keyboard navigation. This should be **off** by default.
- It is possible to perform the following actions through keyboard controls **only**:
  - Navigate between and within lines
  - Select text
  - Trigger code intel hover overlays
  - Interact with code intel hover overlays

## Implementation

1. It should be possible to trigger [this configuration](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@tr/blob-accessibility/-/blob/client/web/src/repo/blob/CodeMirrorBlob.tsx?L33&utm_source=VSCode-2.2.10) via a setting. This will make the blob view navigable with a cursor.
2. We should then be able to trigger `HoverOverlay` through key press. We can listen for key events on the blob and use `window.getSelection` to then get coordinates of where the cursor is and fetch resultant information.
3. We will likely want to autofocus the code intel prompt when it is opened for ease of use, we should ensure that focus is restored properly. We should also ensure that we do not autofocus a disabled button (e.g. Go to def might be disabled)

/cc @taylorsperry @muratsu

**Comment by felixfbecker** [comment]
@umpox I don't know how much testing you did already on this approach, but I don't think this is a requirement nor sufficient for WCAG 2.1, or even a good or acceptable experience for screen reader users to focus different interactive tokens:

https://user-images.githubusercontent.com/10532611/193650387-6b6d8693-18ac-4302-992e-fd2da0c53ece.mp4

Imagine having only the audio here – it is incredibly difficult from just getting read out each character by character to understand what token you are on, especially what interactive token. Notice how it can either read out the whole line, or only character by character, which makes it extremely tedious to try to get to the tokens of interest in the line.

I think https://github.com/sourcegraph/sourcegraph/issues/42421 is the real WCAG requirement and experience a screen reader user needs, being able to read the code parts as normal interactive HTML text (whole words), and getting interactive tokens called out as links with triggerable popovers (`aria-haspopup`), while being able to jump (change focus) between interactive (focusable) tokens quickly. And then getting read out the whole token, not individual characters.
The perfect experience on top of that then would be 2-dimensional focus changes with arrow keys in addition to <kbd>tab</kbd>, but for WCAG 2.1 I think we at a minimum need to make the interactive tokens actual links.

A cursor is a great tool to insert text at specific characters, but we don't have any insertion capabilities, so it makes no sense IMO to add this as a solution to keyboard and screen reader WCAG 2.1 compliance of the interactive blob view.
Put differently: If it was a great tool for general reading and link navigation, we could just make the entire app a read-only `contenteditable` and call accessibility solved. But obviously that would not pass.

This may be a feature that some non-screen-reader users would like for familiarity with editors, but classifying this as a WCAG 2.1 requirement or the solution for WCAG seems wrong. I think [the order in the RFC](https://docs.google.com/document/d/1_y3OhK9V7k1axcvdtroqyjP15dmpvw2J4byaM3FGXW0/edit#heading=h.5ozs19hvrfj3) needs to be swapped, we need tab focus (link elements) for WCAG 2.1, and cursor can be an optional setting post-WCAG, orthogonal to WCAG compliance.

**Comment by eseliger** [comment]
Plus one on that a cursor as in the definition how an editor defines it doesn't really make sense to me for Sourcegraph. Selecting and copying code is IMO not the #1 use case of Sourecgraph (and I would be sad if we are just the "copy paste code platform"), but it's navigating and understanding code.
I personally would much more love to tab through available symbols, being able to skip ALL the noise in between symbols and get right to the next one, then go one level deeper, go to the next thing, repeat.

**Comment by umpox** [comment]
Cursor navigation is never the ideal solution (and should never be on by default) - but it is something we can enable quickly if we _need_ to (essentially we can just update our current hover logic to support cursor logic too). I believe the user experience wouldn't be quite as it seems on that video though (otherwise CM/other editors would be terrible with screen readers), it's a lot better user experience when using screen reader controls to read the next line/word and so on.

I agree with both of you though, https://github.com/sourcegraph/sourcegraph/issues/42421 is a significantly better solution that benefits everyone and _should_ be on by default. I'm currently investigating how feasible this is (until recently I thought we were limited by fetching `CodeIntelligenceRange`, which is too slow for properly tab navigation). I'll leave a more up to date comment on https://github.com/sourcegraph/sourcegraph/issues/42421 with thoughts on where we can go with this. There's still some limitations, but really hoping we can push ahead with this

**Comment by felixfbecker** [comment]
Okay great that we're aligned on this! I also want to point out we don't need to do any work on our part _at all_ to enable users to navigate (and select) code with a cursor, it's built into the browser similar to how tab-focus is. In Chrome and Firefox, all you need to do is press <kbd>F7</kbd> to get a cursor (aka enable "caret browsing") or disable it again. Safari uses the OS-wide support for this.

This is the same method how you would select text on e.g. a Wikipedia article – Wikipedia wouldn't add a built-in cursor with Javascript to enable that.
And as long as tokens are marked up semantically (i.e. use `<a>` elements), this even allows you to navigate too (when you move the caret into a link, it becomes focused automatically, and you can activate it with <kbd>enter</kbd>) – we don't need to add any special keyboard handling to trigger it based on cursor position.

This combined with the fact that cursors don't seem sufficient for WCAG compliance underline to me that it doesn't make sense to invest engineering effort in a custom solution here but all effort should instead be focused on https://github.com/sourcegraph/sourcegraph/issues/42421.

**Comment by lrhacker** [comment]
Given the current state of keyboard navigation work in the blob view (especially with #44698 currently in review) and the previous comments here, I'm going to close this issue.

It sounds like #42421 is what we're looking to do and have already made good progress towards it now. #44698 satisfies most of the requirements (except tab-driven navigation specifically, and instead uses arrow keys and h/j/k/l - the arrow key navigation itself was defined in a separate but similar issue, #43175).
